cip/CIP2015-09-15-STARTS-WITH-and-ENDS-WITH.adoc

```diff
@@ -0,0 +1,144 @@
+* Rely on regular expressions only (difficult for the same reasons as LIKE).
+
+== What others do
+
+=== SQL
+
+SQL uses LIKE - please refer to the link:https://docs.google.com/document/d/1eXhnAS2KmpiAhFTeWssf3s5LrLzUM59B39WJbXa1z_0/edit?usp=sharing[LIKE CIP].
+
+=== SPARQL
+
+SPARQL uses regular expressions only.
+
+=== MongoDB
+
+* If a field is indexed with a text index, a search can be made for documents containing a given term in the field; see link:https://docs.mongodb.org/manual/reference/operator/query/text/[here] for more details.
+* Regex searching is also provided; see link:https://docs.mongodb.org/manual/reference/operator/query/regex/[here] for more details.
+
+=== Elastic (formerly known as Elasticsearch)
+
+* Simple prefix query: this is a low-level query working at the document term level and is not optimised. More details may be found link:https://www.elastic.co/guide/en/elasticsearch/guide/current/prefix-query.html[here].
+* Wildcard term query: this is a low-level, term-based query similar to the prefix query, but allowing for the specification of a pattern. It uses `?` to match any character and `*` to match zero or more characters and is also not optimised. More details may be found link:https://www.elastic.co/guide/en/elasticsearch/guide/current/_wildcard_and_regexp_queries.html[here].
+* Regex searching is also available and is also not optimised. Details may be found link:https://www.elastic.co/guide/en/elasticsearch/guide/current/_wildcard_and_regexp_queries.html[here].
+
+== Benefits to this proposal
+
+(Efficient) string search would be a very frequently-used and important operation that would be supported by implementing this CIP.
+
+== Caveats to this proposal
+
+ * More complex string searches must still use the regular expression search.
+ * Differs from SQL's approach.
+ * More keywords added to the language.
+
+== Appendix
+
+=== Background on case-insensitive search
+
+Case-insensitive string search requires specifying a case folding function for converting the string operands as well as using a suitable equality predicate to compare them. Unicode itself defines three possible cases: lower case, upper case, and title case. Converting to these cases is inherently locale specific though in practice this is often ignored by using the default ("C" or "en") locale. To achieve good results, this is often combined with another locale dependent normalization step. Furthermore notably Java defines a special equality predicate "equalsIgnoreCase" for case insensitive comparison that treats two strings identical if they have the same length and if all of their characters are pairwise equal either directly, or after upcasing both of them, or after downcasing both of them.
```

Comment:
We should replace `case folding function` with `case conversion function` as case folding has a special meaning (which I didn't knew when writing this). Also replace `locale dependent normalization step` with `locale independent normalization step`. Makes more sense this way.